### PR TITLE
ztp: Extend IPsec optional feature to support SNO+1 and MNO

### PR DESCRIPTION
**Pull request by sabinaaledort**
Configure IPsec encryption for external traffic on SNO, SNO+1 and MNO.

### CODE REVIEW

File: ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md [hunk 1]
```diff
@@ -2,57 +2,60 @@
...
-Configure the following values in `ipsec-endpoint-config.yml`:
+#### Configure IPsec
 
-- `$clusterNode` - the IP address or DNS hostname of the cluster node for the IPsec tunnel on the cluster side
-- `$externalHost` - the external host IP address or DNS hostname
-- `$externalAddress` - the IP address or subnet of the external host
+Create an IPsec configuration with an NMState Operator node network configuration policy.
 
-For example:
+Include the policy template file `ipsec-config-policy.yaml` in one of the additional install-time manifests directories.
```
Comment (bradyallenjohnson) [inline ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md:38]: I think we should add a comment here mentioning that the remote-end parameters may need to be modified, namely right and rightsubnet.

**Comment by sabinaaledort** [comment]
cc @yuvalk @lack

File: ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md [hunk 2]
```diff
@@ -2,57 +2,97 @@
...
+* Include `enable-ipsec.yaml` file in one of the additional install-time manifests directories defined in the `extraManifests.searchPaths` field in the SiteConfig file
+More info about the `extraManifests.searchPaths` mechanism can be found in the [GitOps ZTP README](https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/README.md)
+
+Below are two options to configure IPsec encryption for external traffic.
+
+### IPsec configuration with NMState Operator
+
+#### Prerequisites
+
+* NMState Operator installed on the cluster.
```
Comment (yuvalk) [inline ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md:21]: shouldnt we include the subscription for it?
not an expert here, but I believe that's how it's done in other cases,
for example
https://github.com/openshift-kni/cnf-features-deploy/blob/e31eb6cd0b0b2f09472b5de086302656abdb2a48/ztp/source-crs/PtpSubscription.yaml#L5
not sure how to do an optional one though..

Comment (lack) [reply ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md:21]: To make it optional, Add the source-crs for it, but when you add it to the example PGT/PolicyGenerator, comment it out (like [this](https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/example/policygentemplates/common-ranGen.yaml#L59-L67))

File: ztp/gitops-subscriptions/argocd/example/policygentemplates/common-ranGen.yaml [hunk 3]
```diff
@@ -76,6 +76,18 @@ spec:
     # - fileName: OadpOperatorStatus.yaml
     #   policyName: "subscriptions-policy"
     #
+    # NMState operator is used for IPsec configuration with NMState
```
Comment (lack) [inline ztp/gitops-subscriptions/argocd/example/policygentemplates/common-ranGen.yaml:79]: Maybe note that it's only recommended for multi-node deployments?

Comment (yuvalk) [reply ztp/gitops-subscriptions/argocd/example/policygentemplates/common-ranGen.yaml:79]: I believe this is now resolved in the line below

File: ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md [hunk 4]
```diff
@@ -2,57 +2,65 @@
...
+* Include `enable-ipsec.yaml` file in one of the additional install-time manifests directories defined in the `extraManifests.searchPaths` field in the SiteConfig file.
+More info about the `extraManifests.searchPaths` mechanism can be found in the [GitOps ZTP README](https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/README.md)
+* Provide the following certificate files:
+  - `left_server.p12`: The certificate bundle for the IPsec endpoints
+  - `ca.pem`: The certificate authority that you signed your certificates with
 
-* IPsec endpoint certificate in PKCS#12 format and a CA cert in PEM format.
+Below are two options to configure IPsec encryption for external traffic:
+- Using NMState Operator
+- Using a MachineConfig
```
Comment (yuvalk) [inline ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md:18]: I would reword this section so that it is clear 
we support MC for SNO and SNO+1 only
and nmstate operator for MNO only
(ie - it's not really 2 options)

File: ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-config-policy.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                {{- end }}
+                {{- end }}
+                leftid: '%fromcert'
+                leftmodecfgclient: false
+                leftcert: left_server
+                leftrsasigkey: '%cert'
+                right: $externalHost
+                rightid: '%fromcert'
+                rightrsasigkey: '%cert'
+                rightsubnet: $externalAddress
```
Comment (yuvalk) [inline ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-config-policy.yaml:44]: shouldnt this be `rightsubnets`? what happens if customer have multiple networks?

File: ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-config-policy.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                ikev2: insist
+                type: tunnel
+                esp: aes256-sha1
```
Comment (yuvalk) [inline ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-config-policy.yaml:47]: I thought we agreed to use the RHEL default policy.
in any case, we definitely dont want the insecure sha1 here

File: ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md (hunk 4 above)
Comment (sabinaaledort) [reply ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/README.md:18]: I tried to keep this open for the user to choose so I mentioned below our recommendation.  
If you think it's confusing I can change

File: ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-config-policy.yaml (hunk 5 above)
Comment (sabinaaledort) [reply ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-config-policy.yaml:44]: `rightsubnets` field is not supported in nmstate, is there an open RFE?

Comment (sabinaaledort) [reply ztp/gitops-subscriptions/argocd/example/optional-extra-manifest/ipsec/ipsec-config-policy.yaml:44]: There is an open bug https://issues.redhat.com/browse/RHEL-32947

File: ztp/gitops-subscriptions/argocd/example/policygentemplates/common-ranGen.yaml [hunk 7]
```diff
@@ -76,6 +76,19 @@ spec:
...
+    #   policyName: "subscriptions-policy"
+    # - fileName: NMStateSubscriptionOperGroup.yaml
+    #   policyName: "subscriptions-policy"
+    # - fileName: NMStateSubscription.yaml
+    #   policyName: "subscriptions-policy"
+    # - fileName: NMStateOperatorStatus.yaml
+    #   policyName: "subscriptions-policy"
+    # - fileName: NMState.yaml
+    #   policyName: "subscriptions-policy"
+    #
```
Comment (lack) [inline ztp/gitops-subscriptions/argocd/example/policygentemplates/common-ranGen.yaml:91]: Let's move this to `group-du-3node-ranGen.yaml` and `group-du-standard-ranGen.yaml` so there's no mistaking where this operator is allowed.
